src/ui/libs/DatalensChartkit/ChartKit/components/Widget/components/Table/utils/misc.ts

```diff
@@ -356,6 +362,170 @@ function extractCellActionParams(args: {cell: TableCell; head?: TableHead}) {
     return {};
 }
 
+function setMapValue(map: ValuesMap, value: string, hash: string) {
+    if (typeof map[value] === 'undefined') {
+        // eslint-disable-next-line no-param-reassign
+        map[value] = {value: 0, hashes: []};
+    }
+
+    if (map[value].hashes.indexOf(hash) === -1) {
+        map[value].hashes.push(hash);
+    }
+
+    // eslint-disable-next-line no-param-reassign
+    map[value].value = map[value].value + 1;
+}
+
+function getValuesMap(args: {selectedRows: TableRow[]; head?: TableHead[]}) {
+    const {selectedRows, head} = args;
+
+    return selectedRows.reduce<ValuesMap>((acc, row) => {
+        if (!('cells' in row)) {
+            return acc;
+        }
+
+        const rowHash = Object.values(row.cells).reduce<string>((acc, cell, i) => {
+            const cellParams = extractCellActionParams({cell, head: head?.[i]});
+            return acc + Object.values(cellParams).join();
+        }, '');
+
+        row.cells.forEach((cell, i) => {
+            const cellParams = extractCellActionParams({cell, head: head?.[i]});
+            Object.values(cellParams).forEach((cellValue) => {
+                if (Array.isArray(cellValue)) {
+                    cellValue.forEach((value) => {
+                        setMapValue(acc, value, rowHash);
+                    });
+                } else {
+                    setMapValue(acc, cellValue, rowHash);
+                }
+            });
+        });
+
+        return acc;
+    }, {});
+}
+
+function shouldRemoveValue(map: ValuesMap, key: string) {
+    return (
+        // Value is contained only in one row
+        map[key]?.value === 1 ||
+        // Values are contained in completely identical rows
+        (map[key]?.value > 1 && map[key].hashes.length === 1)
+    );
+}
+
+function mergeStringParamsByRowDeselecting(args: {
+    current: StringParams;
+    row: StringParams;
+    selectedRows: TableRow[];
+    head?: TableHead[];
+}) {
+    const {current, row, selectedRows, head} = args;
+    return Object.keys(current).reduce<StringParams>((acc, key) => {
+        acc[key] = current[key];
+        const valuesMap = getValuesMap({selectedRows, head});
+        const hasSelectedRows = Boolean(selectedRows.length);
```

Comment:
Is it necessary to calculate the `valuesMap`  for each `current[key]`?
Perhaps you can move it higher?